simulation_test.go

```diff
@@ -0,0 +1,221 @@
+	data, err := ioutil.ReadFile(itfFileName)
+	if err != nil {
+		t.Errorf("Error opening file: %v", err)
+	}
+	states := gjson.GetBytes(data, "states").Array()
+
+	// iterate over all states of the test run
+	for stateIdx, state := range states {
+		t.Logf("State %v\n", stateIdx)
+		// because the test run consists of multiple steps to generate the tree and the proof, test
+		// only the states marked "final"
+		if state.Get("state_v").String() == "final" {
+			modelNamespace := int(state.Get("namespace_v").Int())
+			modelLeaves := getLeavesFromState(state, nidSize, "leaves_v")
+			modelProof := state.Get("proof_v")
+			corrupted := state.Get("corrupted").Bool()
+			t.Logf("Obtained state data:\n\tState namespace: %v\n\tleaves: %v\n\tcorrupted: %v\n",
+				modelNamespace, modelLeaves, corrupted)
+
+			tree = nmt.New(sha256.New(), nmt.NamespaceIDSize(nidSize))
+			for _, leaf := range modelLeaves {
+				err := tree.Push(leaf)
+				if err != nil {
+					t.Errorf("Error on push: %v", err)
+				}
+			}
+
+			namespaceBytes := intToBytes(modelNamespace, nidSize)
+
+			proof, err := tree.ProveNamespace(namespaceBytes)
+			if err != nil {
+				t.Errorf("Error on prove: %v", err)
+			}
+			var new_hashes [][]byte
+			t.Logf("Proof: %v\n", proof)
+
+			var corruption_type string
+			if !corrupted {
+				proofMatching := checkProof(modelProof, proof, nidSize)
+				if !proofMatching {
+					t.Errorf("Proofs do not match. Expected: %v, got: %v", state, proof)
+				}
+			} else {
+				// CORRUPTED: now we are modifying the proof to be corrupted here as well
+				corruption_type = state.Get("corruption_type").String()
+				t.Log("Corruption type: ", corruption_type)
+				if corruption_type == "supporting_hashes" {
+					for idx := 0; idx < len(state.Get("corruption_diff.changed_indices").Array()); idx++ {
+						new_hashes = append(new_hashes, proof.Nodes()[idx])
+					}
+				} else {
+					new_hashes = proof.Nodes()
+				}
+				proof = nmt.NewInclusionProof(
+					int(modelProof.Get("start").Int()),
+					int(modelProof.Get("end").Int()),
+					new_hashes,
+					true,
+				)
+				t.Logf("Corrupted proof: %v\n", proof)
+
+			}
+
+			root, _ := tree.Root()
+
+			dataToVerify := modelLeaves[proof.Start():proof.End()]
+
+			t.Logf("Verifying. Root: %v, Namespace: %x, Leaves: %v, Proof: %v\n", root, namespaceBytes, dataToVerify, proof)
+
+			successVerification := proof.VerifyNamespace(sha256.New(), namespaceBytes, dataToVerify, root)
+			t.Logf("Verification result: %v\n", successVerification)
+
+			if successVerification == corrupted {
+				successVerification := proof.VerifyNamespace(sha256.New(), namespaceBytes, dataToVerify, root)
+				t.Errorf("Mismatch! Data corrupted = %v, verification result = %v", corrupted, successVerification)
+			}
+
+		} else { // these are the intermediate states, at which we are not checking anything
+			continue
+		}
+
+	}
+}
+
+func TestFromScenario(t *testing.T) {
+	itfFileName := "formal_spec/ITF_files/scenario.json"
```

Comment:
[question] While quint is not specifically designed for generating unit test data, based on the current implementation, is it necessary for each unit test to have its own separate JSON file containing the test data?

Reply:
> If you are asking about different data for the same unit test (as they are currently written in go test files): it is not necessary, but at the moment there is no single quint command that would produce one json with multiple tests (multiple invariant violations).

I see, this was what I was wondering about. Thanks.